server/testkit/tests/https_extractors.rs

```diff
@@ -0,0 +1,91 @@
+use kanidm_client::{KanidmClient, APPLICATION_JSON};
+use kanidm_proto::v1::{AuthIssueSession, AuthRequest, AuthStep};
+use reqwest::header::CONTENT_TYPE;
+
+#[kanidmd_testkit::test(trust_x_forward_for = false)]
+async fn dont_trust_xff_send_header(rsclient: KanidmClient) {
+    let addr = rsclient.get_url();
+    let client = reqwest::ClientBuilder::new()
+        .danger_accept_invalid_certs(true)
+        .build()
+        .unwrap();
+
+    let auth_init = AuthRequest {
+        step: AuthStep::Init2 {
+            username: "anonymous".to_string(),
+            issue: AuthIssueSession::Token,
+        },
+    };
+
+    let response = client
+        .post([&addr, "/v1/auth"].concat())
+        .header(
+            "X-Forwarded-For",
+            "An invalid header that will get through!!!",
+        )
+        .header(CONTENT_TYPE, APPLICATION_JSON)
+        .body(serde_json::to_string(&auth_init).unwrap())
+        .send()
+        .await
+        .unwrap();
+    assert!(dbg!(response).status() == 200);
+}
+
+#[kanidmd_testkit::test(trust_x_forward_for = true)]
+async fn dont_trust_xff_dont_send_header(rsclient: KanidmClient) {
+    // Now login as anonymous
+    let res = rsclient.auth_anonymous().await;
+    assert!(res.is_ok());
+
+    // Now do a whoami.
+    let e = rsclient
+        .whoami()
+        .await
+        .expect("Unable to call whoami")
+        .expect("No entry matching self returned");
+    assert!(e.attrs.get("spn") == Some(&vec!["anonymous@localhost".to_string()]));
+}
+
+#[kanidmd_testkit::test(trust_x_forward_for = true)]
+async fn trust_xff_send_header(rsclient: KanidmClient) {
+    let addr = rsclient.get_url();
+    let client = reqwest::ClientBuilder::new()
+        .danger_accept_invalid_certs(true)
+        .build()
+        .unwrap();
+
+    let auth_init = AuthRequest {
+        step: AuthStep::Init2 {
+            username: "anonymous".to_string(),
+            issue: AuthIssueSession::Token,
+        },
+    };
+
+    let response = client
+        .post([&addr, "/v1/auth"].concat())
+        .header(
+            "X-Forwarded-For",
+            "a VERY much invalid header that WON'T get through!!",
+        )
+        .header(CONTENT_TYPE, APPLICATION_JSON)
+        .body(serde_json::to_string(&auth_init).unwrap())
+        .send()
+        .await
+        .unwrap();
+    assert!(dbg!(response).status() == 400);
```

Comment:
we should also be testing the "trust header, send valid header" cases with
- single IPv4 and IPv6 addresses
- combined lists of the above to test the "leftmost" address. ie, (IPv4, IPv4), (IPv4, IPv6), (IPv6, IPv6), (IPv6, IPv4)

Reply:
Sounds good :). btw I've noticed we reject requests only if the leftmost address is invalid, is this expected?

Reply:
Since we're only trusting the left-most address, then ... it seems OK? We probably *should* check all of them to ensure it's a "valid" statement.

Reply:
We might have to end up logging the other right side addresses too, so probably we should check these :)